mailbox/api/src/main/java/org/apache/james/mailbox/events/EventBus.java

```diff
@@ -56,7 +61,28 @@ static String timerName(MailboxListener mailboxListener) {
         return dispatch(event, ImmutableSet.of(key));
     }
 
-    default Registration register(MailboxListener.GroupMailboxListener groupMailboxListener) {
-        return register(groupMailboxListener, groupMailboxListener.getDefaultGroup());
+    /**
+     * @throws GroupsAlreadyRegistered when any initialize method had already been called
```

Comment:
runtime error to enforce initialization is not really dev friendly. Can't we ensure listeners are registered at construction time instead?

Reply:
We are limited by guice ability: the list of listener is configured.

In short: I have no idea of how to do that.

And even if I succeed to do that with Guice, I still have to do it with Spring.

Sorry, I know it is not ideal, but I have really no idea of how to contribute such a thing.

Reply:
I suppose you refers to `MailboxListenersLoaderImpl` implementation.
So this class takes a configuration, loads configured listeners and put it as bus listeners when `configure` is called.
It's basically a class having side-effect on another class at startup.

What we can do is:

* create a InitializedEventBus type that takes the list of instantiated listeners in its constructor
* Let MailboxListenersLoaderImpl be a provider of this type InitializedEventBus (it would take an EventBus as input)
* stop using EventBus directly in the code an use InitializedEventBus instead

Would it work?

Reply:
Would do on monday. Could work, but it is complicated.

Reply:
It's a recursive dependency nightmare.

Reply:
I did provide a proof a concept for such a topic. https://github.com/linagora/james-project/pull/3292 provides several strategies for recursive dependencies issues. I would like to have a deep talk about it.